internal/core/src/storage/parquet_c.cpp

```diff
@@ -0,0 +1,556 @@
+            st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+            st.error_msg = ErrorMsg("incorrect data type");
+            return st;
+        }
+        p->builder = std::make_shared<arrow::FixedSizeBinaryBuilder>(arrow::fixed_size_binary(dimension / 8));
+        p->schema = arrow::schema({arrow::field("val", arrow::fixed_size_binary(dimension / 8))});
+        p->dimension = dimension;
+    } else if (p->dimension != dimension) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("dimension changed");
+        return st;
+    }
+    auto builder = std::dynamic_pointer_cast<arrow::FixedSizeBinaryBuilder>(p->builder);
+    if (builder == nullptr) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("incorrect data type");
+        return st;
+    }
+    if (p->output != nullptr) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("payload has finished");
+        return st;
+    }
+    auto ast = builder->AppendValues(values, length);
+    if (!ast.ok()) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg(ast.message());
+        return st;
+    }
+    p->rows += length;
+    return st;
+}
+
+extern "C" CStatus
+AddFloatVectorToPayload(CPayloadWriter payloadWriter, float* values, int dimension, int length) {
+    CStatus st;
+    st.error_code = static_cast<int>(ErrorCode::SUCCESS);
+    st.error_msg = nullptr;
+    if (length <= 0)
+        return st;
+
+    auto p = reinterpret_cast<wrapper::PayloadWriter*>(payloadWriter);
+    if (p->dimension == wrapper::EMPTY_DIMENSION) {
+        if (p->builder != nullptr) {
+            st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+            st.error_msg = ErrorMsg("incorrect data type");
+            return st;
+        }
+        p->builder =
+            std::make_shared<arrow::FixedSizeBinaryBuilder>(arrow::fixed_size_binary(dimension * sizeof(float)));
+        p->schema = arrow::schema({arrow::field("val", arrow::fixed_size_binary(dimension * sizeof(float)))});
+        p->dimension = dimension;
+    } else if (p->dimension != dimension) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("dimension changed");
+        return st;
+    }
+    auto builder = std::dynamic_pointer_cast<arrow::FixedSizeBinaryBuilder>(p->builder);
+    if (builder == nullptr) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("incorrect data type");
+        return st;
+    }
+    if (p->output != nullptr) {
+        st.error_code = static_cast<int>(ErrorCode::UNEXPECTED_ERROR);
+        st.error_msg = ErrorMsg("payload has finished");
+        return st;
+    }
+    auto ast = builder->AppendValues(reinterpret_cast<const uint8_t*>(values), length);
```

Comment:
Maybe `static_cast` is more reasonable here than `reinterpret_cast`.

Reply:
these files are moved from internal/storage/cwrapper, and formatted code.